ComponentKit/Utilities/CKArrayControllerChangesetVerification.mm

```diff
@@ -0,0 +1,154 @@
+      invalidChangeFound = YES;
+      *stop = YES;
+    }
+  }, nil, nil, nil);
+
+  if (invalidChangeFound) {
+    return CKBadChangesetOperationTypeUpdate;
+  }
+
+  /** 
+   Removed rows
+   Note we need to use the dictionary of index sets so that we have an index set for each section.
+   For discussion about why we need index sets at all, see the 'removed sections' comment below
+   */
+  NSMutableDictionary<NSNumber *, NSMutableIndexSet *> *rowsToRemove = [NSMutableDictionary new];
+  changeset.items.enumerateItems(nil, ^(NSInteger section, NSInteger index, BOOL *stop) {
+    if (section >= [currentSectionCounts count] || section < 0 || index >= [currentSectionCounts[section] integerValue]) {
+      invalidChangeFound = YES;
+      *stop = YES;
+    } else {
+      if (!rowsToRemove[@(section)]) {
+        rowsToRemove[@(section)] = [NSMutableIndexSet indexSet];
+      }
+
+      [rowsToRemove[@(section)] addIndex:index];
+    }
+  }, nil, nil);
+
+  if (invalidChangeFound) {
+    return CKBadChangesetOperationTypeRemoveRow;
+  } else {
+    [rowsToRemove enumerateKeysAndObjectsUsingBlock:^(NSNumber * _Nonnull section, NSMutableIndexSet * _Nonnull indexSet, BOOL * _Nonnull stop) {
+      currentSectionCounts[[section integerValue]] = @([currentSectionCounts[[section integerValue]] integerValue] - [indexSet count]);
+    }];
+  }
+
+  /** 
+   Removed sections
+   
+   We need to keep an index set to track what sections to remove instead of removing each inline because we cannot
+   guarantee the order of the removals, so for example:
+   
+   Let's say we have 3 sections, and we want to remove the second and third section. If we remove the second section first,
+   we'll end up trying to remove the third section when there's only 2 sections left, and we'll get a false negative.
+   */
+  NSMutableIndexSet *sectionsToRemove = [NSMutableIndexSet indexSet];
+  for (NSInteger removal : changeset.sections.removals()) {
+    if (removal >= [currentSectionCounts count]) {
+      return CKBadChangesetOperationTypeRemoveSection;
+    } else {
+      [sectionsToRemove addIndex:removal];
+    }
+  }
+  [currentSectionCounts removeObjectsAtIndexes:sectionsToRemove];
+
+  // Inserted sections
+  for (NSInteger insertion : changeset.sections.insertions()) {
+    // Strictly greater than, since you can always insert at the end
+    if (insertion > [currentSectionCounts count]) {
+      return CKBadChangesetOperationTypeInsertSection;
+    } else {
+      [currentSectionCounts insertObject:@0 atIndex:insertion];
+    }
+  }
+
+  // Moved sections
+  for (std::pair<NSInteger, NSInteger> move : changeset.sections.moves()) {
+    NSInteger origin = move.first;
+    NSInteger destination = move.second;
+
+    if (origin >= [currentSectionCounts count] || destination >= [currentSectionCounts count]) {
+      return CKBadChangesetOperationTypeMoveSection;
+    } else {
+      NSNumber *originValue = currentSectionCounts[origin];
```

Comment:
`originValue` is a bit unclear. If these are section counts then how about `originSectionCount`. Same goes for `destinationValue`.